pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/TopicPoliciesSystemTopicClient.java

```diff
@@ -85,42 +85,48 @@ private TopicPolicyWriter(Producer<PulsarEvent> producer, SystemTopicClient<Puls
         }
 
         @Override
-        public MessageId write(PulsarEvent event) throws PulsarClientException {
-            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(getEventKey(event)).value(event);
+        public MessageId write(PulsarEvent event, String key) throws PulsarClientException {
+            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(key).value(event);
             setReplicateCluster(event, builder);
             return builder.send();
         }
 
         @Override
-        public CompletableFuture<MessageId> writeAsync(PulsarEvent event) {
-            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(getEventKey(event)).value(event);
+        public CompletableFuture<MessageId> writeAsync(PulsarEvent event, String key) {
+            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(key).value(event);
             setReplicateCluster(event, builder);
             return builder.sendAsync();
         }
 
         @Override
-        public MessageId delete(PulsarEvent event) throws PulsarClientException {
-            validateActionType(event);
-            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(getEventKey(event)).value(null);
+        public MessageId delete(PulsarEvent event, String key) throws PulsarClientException {
+            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(key).value(null);
             setReplicateCluster(event, builder);
             return builder.send();
         }
 
         @Override
-        public CompletableFuture<MessageId> deleteAsync(PulsarEvent event) {
+        public CompletableFuture<MessageId> deleteAsync(PulsarEvent event, String key) {
             validateActionType(event);
-            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(getEventKey(event)).value(null);
+            TypedMessageBuilder<PulsarEvent> builder = producer.newMessage().key(key).value(null);
             setReplicateCluster(event, builder);
             return builder.sendAsync();
         }
 
-        private String getEventKey(PulsarEvent event) {
+        public static String getEventKey(PulsarEvent event) {
             return TopicName.get(event.getTopicPoliciesEvent().getDomain(),
                 event.getTopicPoliciesEvent().getTenant(),
                 event.getTopicPoliciesEvent().getNamespace(),
                 event.getTopicPoliciesEvent().getTopic()).toString();
         }
 
+        public static String getEventKey(TopicName topicName) {
+            return TopicName.get(topicName.getDomain().toString(),
+                    topicName.getTenant(),
+                    topicName.getNamespace(),
+                    TopicName.get(topicName.getPartitionedTopicName()).getLocalName()).toString();
+        }
```

Comment:
I think the caller should provide the `getEventKey` implementation, not the `TopicPoliciesSystemTopicClient`? Please keep the responsibilities clear.